include/PreferenceList.h

```diff
@@ -27,6 +28,8 @@ class PreferenceList {
 private:
     RankType cur_rank_;        // the current unused rank
     ContainerType pref_list_;  // strictly ordered preference list
+    std::unordered_map<RankType, std::vector<PrefListElement>> tie_map_; //map to store when a particular rank has ties 
+    std::unordered_map<RankType, bool> is_tied_;
```

Comment:
This map is not needed. You can probe tie_map_ to check if a RankType is present as key. This will tell you if this rank is tied.